osu.Game.Tests/Visual/Ranking/TestSceneResultsScreen.cs

```diff
@@ -411,84 +423,77 @@ protected override void LoadComplete()
+        }
+    }
+}
 
-            public bool FetchCompleted { get; private set; }
+public class TestScores
+{
+    public IBindableList<ScoreInfo> Leaderboard => Scores;
+    protected BindableList<ScoreInfo> Scores = new BindableList<ScoreInfo>();
 
-            public DelayedFetchResultsScreen(ScoreInfo score, Task fetchWaitTask = null)
-                : base(score)
-            {
-                this.fetchWaitTask = fetchWaitTask ?? Task.CompletedTask;
-            }
+    public TestScores()
+    {
+        FetchScores();
+    }
 
-            protected override APIRequest FetchScores(Action<IEnumerable<ScoreInfo>> scoresCallback)
-            {
-                Task.Run(async () =>
-                {
-                    await fetchWaitTask;
+    public virtual APIRequest FetchScores()
+    {
+        var scores = new List<ScoreInfo>();
+
+        for (int i = 0; i < 20; i++)
+        {
+            var score = TestResources.CreateTestScoreInfo();
+            score.TotalScore += 10 - i;
+            score.HasOnlineReplay = true;
+            scores.Add(score);
+        }
 
-                    var scores = new List<ScoreInfo>();
+        Scores.Clear();
+        Scores.AddRange(scores);
 
-                    for (int i = 0; i < 20; i++)
-                    {
-                        var score = TestResources.CreateTestScoreInfo();
-                        score.TotalScore += 10 - i;
-                        scores.Add(score);
-                    }
+        return null;
+    }
+}
 
-                    scoresCallback?.Invoke(scores);
+public class DelayedTestScores : TestScores
+{
+    private readonly Task fetchWaitTask;
 
-                    Schedule(() => FetchCompleted = true);
-                });
+    public bool FetchCompleted { get; private set; }
 
-                return null;
-            }
-        }
+    public DelayedTestScores(Task fetchWaitTask = null)
+    {
+        this.fetchWaitTask = fetchWaitTask ?? Task.CompletedTask;
+        FetchScores();
+    }
 
-        private partial class UnrankedSoloResultsScreen : SoloResultsScreen
+    public override APIRequest FetchScores()
+    {
+        Task.Run(async () =>
         {
-            public HotkeyRetryOverlay RetryOverlay;
+            await fetchWaitTask;
 
-            public UnrankedSoloResultsScreen(ScoreInfo score)
-                : base(score)
+            var scores = new List<ScoreInfo>();
+
+            for (int i = 0; i < 20; i++)
             {
-                AllowRetry = true;
-                Score!.BeatmapInfo!.OnlineID = 0;
-                Score.BeatmapInfo.Status = BeatmapOnlineStatus.Pending;
+                var score = TestResources.CreateTestScoreInfo();
+                score.TotalScore += 10 - i;
+                scores.Add(score);
             }
 
-            protected override void LoadComplete()
-            {
-                base.LoadComplete();
+            Scores.Clear();
+            Scores.AddRange(scores);
```

Comment:
Without looking at the rest of this diff, this is pretty conclusively unsafe with the rest of it. `BindableList` is not an atomically readable collection. As in, you can't just mutate it from thread A and safely enumerate it in the callback it on thread B (that being the update thread in `SoloResultsScreen`).

The only thing that's safe to access in a multi-threaded context is the stuff in the `NotifyCollectionChangedEventArgs` callback, so if you were relying on `BindableList` for thread safety, then that's not gonna work. Either use the callbacks, or switch to `Bindable<ImmutableList<T>>` or something.